### ISSUE TITLE
M-FAC-like pruner crashes due to the GPUtil not seeing any GPUs

### ISSUE THREAD

**Issue by Godofnothing** [neuralmagic/sparseml#645]
I am experimenting with `M-FAC` like `BaseGradualPruningModifier` and what I've Implemented is the class inheriting
from either `MFACPruningModifier` and `ACDCPruningModifier` which I've called `ACDCMFACPruningModifier`.

```
@PyTorchModifierYAML()
class ACDCMFACPruningModifier(ACDCBasePruningModifier, MFACPruningModifier):
   ....
```

```
I init the Modifier via the SparseML recipe:

  - !ACDCMFACPruningModifier
    compression_sparsity: 0.75
    start_epoch: 150
    end_epoch: 500
    update_frequency: 10
    params: ['re:(.*blocks.*.attn.qkv.weight)|(.*blocks.*.attn.proj.weight)|(.*blocks.*.mlp.fc.*weight)'] 
    mask_type: unstructured
    global_sparsity : True
    num_grads: 256
    fisher_block_size: 64
```

And for the long run, as in this recipe at the start of the compression stage training crashes on this step:

```
2022-03-27 21:37:34 sparseml.pytorch.sparsification.pruning.modifier_pruning_mfac INFO     Using Small Block Fast Fisher Inverse Implementation 
Using Small Block Fast Fisher Inverse Implementation 
...
File "/nfs/scistore14/alistgrp/dkuznede/miniconda3/envs/deep_learning/lib/python3.9/site-packages/sparseml/pytorch/sparsification/pruning/modifier_pruning_mfac.py", line 1506, in _get_free_gpu_memory 
    return [gpus_all[idx].memoryFree for idx in gpu_idx] 
  File "/nfs/scistore14/alistgrp/dkuznede/miniconda3/envs/deep_learning/lib/python3.9/site-packages/sparseml/pytorch/sparsification/pruning/modifier_pruning_mfac.py", line 1506, in <listcomp> 
    return [gpus_all[idx].memoryFree for idx in gpu_idx] 
IndexError: list index out of range 
```

I've delved into the issue and found out, that at this step:
`gpus_all = GPUtil.getGPUs()`
returns empty list. 

I've rerun the experiment with a short run with a few epochs.

    start_epoch: 2
    end_epoch: 8
    update_frequency: 1

There were no any issues and the GPUtil returned the list of all GPUs available. Training finished normally. 

The SparseML recipe is instantiated in the standard way:

    manager = ScheduledModifierManager.from_yaml(args.sparseml_recipe)
    # wrap optimizer   
    optimizer = manager.modify(model, optimizer, steps_per_epoch=len(loader_train))

Seems like it's some a hardware-like issue underhood. What could you suggest to look into?

 I am working with the nightly release of SparseML.

Version of packages installed
---
sparseml-nightly              0.12.0.20220326 
cudatoolkit               11.3.1               h2bc3f7f_2
GPUtil                        1.4.0
torch                         1.10.2

System
---
GPU - `RTX3090`
OS - `Debian GNU/Linux 10 (buster)`

**Comment by jeanniefinks** [comment]
Thank you @Godofnothing for the issue and details.
We will take a closer look and get back to you. 
Best, Jeannie / Neural Magic

**Comment by markurtz** [comment]
Hi @Godofnothing, the issue that's happening is the GPU utils package we're using is failing to return the proper info later in the runs. We haven't seen this issue before and are unsure of why this might be popping up. To enable a quick remedy we've put up the following diff to attempt to cache those results: https://github.com/neuralmagic/sparseml/pull/650. This will be landing shortly on main and will be pushed up to the next nightly if you'd like to see if that remedies the issues you're seeing. 

Thanks,
Mark

**Comment by bfineran** [comment]
Closed due to #650 landing which adds the proposed fix to the issue. Please reopen if any further issues are noticed.

**Comment by Godofnothing** [comment]
@bfineran , @KSGulin thanks all for help!
